File: tools/adapters/ocaml/src/trustforge/trustforge.ml
(** TrustForge shared HTTP client for OCaml.

    Speaks [POST /v1/decide] against a local tf-daemon and returns a
    [decide_response]. Uses [Cohttp_lwt_unix] for the HTTP transport
    and [Yojson] for JSON. *)

type mode = Enforce | Observe_only

type config = {
  daemon_url : string;
  admin_token : string option;
  mode : mode;
  timeout_ms : int;
}

let default_config = {
  daemon_url = "http://127.0.0.1:8787";
  admin_token = None;
  mode = Enforce;
  timeout_ms = 5000;
}

type decision =
  | Allow
  | Deny
  | Approval_required
  | Escalate
  | Log_only
  | Unknown of string

let decision_of_string = function
  | "allow" -> Allow
  | "deny" -> Deny
  | "approval-required" -> Approval_required
  | "escalate" -> Escalate
  | "log-only" -> Log_only
  | other -> Unknown other

let string_of_decision = function
  | Allow -> "allow"
  | Deny -> "deny"
  | Approval_required -> "approval-required"
  | Escalate -> "escalate"
  | Log_only -> "log-only"
  | Unknown s -> s

type decide_request = {
  action : string;
  host_token : string option;
  host_token_kind : string option;
  target : string option;
  trace_id : string option;
}

let make_request ?host_token ?host_token_kind ?target ?trace_id action =
  { action; host_token; host_token_kind; target; trace_id }

type decide_response = {
  decision : decision;
  reason : string;
  proof_id : string;
  approval_id : string option;
  danger_tags : string list;
}

type error =
  | Daemon_unavailable of string
  | Daemon_rejected of int * string
  | Invalid_response of string

exception Trustforge_error of error

let error_to_string = function
  | Daemon_unavailable s -> Printf.sprintf "daemon-unavailable: %s" s
  | Daemon_rejected (c, s) -> Printf.sprintf "daemon-rejected[%d]: %s" c s
  | Invalid_response s -> Printf.sprintf "invalid-response: %s" s

(** Encode a [decide_request] as a JSON string body. *)
let encode_request_body req : string =
  let pairs = ref [ ("action", `String req.action) ] in
  (match req.host_token with
   | Some v -> pairs := !pairs @ [ ("host_token", `String v) ]
   | None -> ());
  (match req.host_token_kind with
   | Some v -> pairs := !pairs @ [ ("host_token_kind", `String v) ]
   | None -> ());
  (match req.target with
   | Some v -> pairs := !pairs @ [ ("target", `String v) ]
   | None -> ());
  (match req.trace_id with
   | Some v -> pairs := !pairs @ [ ("trace_id", `String v) ]
   | None -> ());
  Yojson.Safe.to_string (`Assoc !pairs)

let assoc_string key obj =
  match List.assoc_opt key obj with
  | Some (`String s) -> Some s
  | _ -> None

let assoc_string_list key obj =
  match List.assoc_opt key obj with
  | Some (`List xs) ->
      List.filter_map (function `String s -> Some s | _ -> None) xs
  | _ -> []

(** Parse a JSON response body into a [decide_response]. *)
let parse_response_body body : (decide_response, error) result =
  match Yojson.Safe.from_string body with
  | exception (Yojson.Json_error msg) -> Error (Invalid_response msg)
  | `Assoc obj ->
      let decision_str =
        Option.value (assoc_string "decision" obj) ~default:"unknown"
      in
      let reason = Option.value (assoc_string "reason" obj) ~default:"" in
      let proof_id = Option.value (assoc_string "proof_id" obj) ~default:"" in
      let approval_id = assoc_string "approval_id" obj in
      let danger_tags = assoc_string_list "danger_tags" obj in
      Ok {
        decision = decision_of_string decision_str;
        reason;
        proof_id;
        approval_id;
        danger_tags;
      }
  | _ -> Error (Invalid_response "expected JSON object")

(** Pull a Bearer token out of an Authorization header value. *)
let extract_bearer h =
  if String.length h <= 7 then None
  else
    let prefix = String.sub h 0 7 in
    let lower = String.lowercase_ascii prefix in
    if lower <> "bearer " then None
    else
      let raw = String.sub h 7 (String.length h - 7) in
      let trimmed = String.trim raw in
      if trimmed = "" then None else Some trimmed

(** Configurable client. Stores the [config] and reuses an HTTP context. *)
type t = { config : config }

let create config = { config }

let config_of t = t.config

(** Perform a [/v1/decide] call. *)
let decide (t : t) (req : decide_request) :
    (decide_response, error) result Lwt.t =
  let open Lwt.Infix in
  let body = encode_request_body req in
  let url = t.config.daemon_url ^ "/v1/decide" in
  let uri = Uri.of_string url in
  let base_headers =
    Cohttp.Header.init_with "content-type" "application/json"
  in
  let headers =
    match t.config.admin_token with
    | Some tok -> Cohttp.Header.add base_headers "authorization" ("Bearer " ^ tok)
    | None -> base_headers
  in
  Lwt.catch
    (fun () ->
      Cohttp_lwt_unix.Client.post
        ~headers
        ~body:(Cohttp_lwt.Body.of_string body)
        uri
      >>= fun (resp, body) ->
      let code = Cohttp.Code.code_of_status (Cohttp.Response.status resp) in
      Cohttp_lwt.Body.to_string body >>= fun body_str ->
      if code >= 500 then
        Lwt.return (Error (Daemon_unavailable (Printf.sprintf "status %d" code)))
      else if code >= 400 then
        Lwt.return (Error (Daemon_rejected (code, body_str)))
      else Lwt.return (parse_response_body body_str))
    (fun exn ->
      Lwt.return (Error (Daemon_unavailable (Printexc.to_string exn))))

File: tools/adapters/haskell/trustforge/src/Trustforge.hs
{-# LANGUAGE OverloadedStrings    #-}
{-# LANGUAGE DeriveGeneric        #-}
{-# LANGUAGE RecordWildCards      #-}
{-# LANGUAGE ScopedTypeVariables  #-}

-- |
-- Module      : Trustforge
--
-- Shared client for the TrustForge tf-daemon HTTP endpoint
-- @POST /v1/decide@. Designed to be used by Haskell web framework
-- adapters (WAI, Yesod, Servant, ...).
module Trustforge
  ( -- * Configuration
    Config (..)
  , defaultConfig
  , Mode (..)
    -- * Requests / responses
  , DecideRequest (..)
  , DecideResponse (..)
  , Decision (..)
  , parseDecision
  , decisionToText
  , emptyRequest
    -- * High-level API
  , Client
  , clientConfig
  , newClient
  , decide
    -- * Errors
  , TrustforgeError (..)
    -- * Helpers
  , extractBearer
  , encodeRequestBody
  , parseResponseBody
  ) where

import qualified Control.Exception          as E
import qualified Data.Aeson                 as A
import           Data.Aeson                 ((.:?), (.!=))
import qualified Data.Aeson.Types           as AT
import qualified Data.ByteString            as BS
import qualified Data.ByteString.Char8      as BC
import qualified Data.ByteString.Lazy       as LBS
import qualified Data.CaseInsensitive       as CI
import qualified Data.Text                  as T
import qualified Data.Text.Encoding         as TE
import           GHC.Generics               (Generic)
import qualified Network.HTTP.Client        as H
import qualified Network.HTTP.Types.Method  as HM
import qualified Network.HTTP.Types.Status  as HS

-- | Enforcement mode. 'ObserveOnly' lets requests through when the
-- daemon is unreachable (logging only); 'Enforce' fails closed.
data Mode = Enforce | ObserveOnly deriving (Eq, Show)

-- | TrustForge client configuration.
data Config = Config
  { cfgDaemonUrl  :: T.Text
  , cfgAdminToken :: Maybe T.Text
  , cfgMode       :: Mode
  , cfgTimeoutMs  :: Int
  } deriving (Eq, Show)

defaultConfig :: Config
defaultConfig = Config
  { cfgDaemonUrl  = "http://127.0.0.1:8787"
  , cfgAdminToken = Nothing
  , cfgMode       = Enforce
  , cfgTimeoutMs  = 5000
  }

-- | TrustForge decision returned by the daemon.
data Decision
  = Allow
  | Deny
  | ApprovalRequired
  | Escalate
  | LogOnly
  | UnknownDecision T.Text
  deriving (Eq, Show, Generic)

decisionToText :: Decision -> T.Text
decisionToText Allow              = "allow"
decisionToText Deny               = "deny"
decisionToText ApprovalRequired   = "approval-required"
decisionToText Escalate           = "escalate"
decisionToText LogOnly            = "log-only"
decisionToText (UnknownDecision t)= t

parseDecision :: T.Text -> Decision
parseDecision t = case t of
  "allow"             -> Allow
  "deny"              -> Deny
  "approval-required" -> ApprovalRequired
  "escalate"          -> Escalate
  "log-only"          -> LogOnly
  other               -> UnknownDecision other

data DecideRequest = DecideRequest
  { reqAction        :: T.Text
  , reqHostToken     :: Maybe T.Text
  , reqHostTokenKind :: Maybe T.Text
  , reqTarget        :: Maybe T.Text
  , reqTraceId       :: Maybe T.Text
  } deriving (Eq, Show)

emptyRequest :: T.Text -> DecideRequest
emptyRequest a = DecideRequest a Nothing Nothing Nothing Nothing

instance A.ToJSON DecideRequest where
  toJSON DecideRequest{..} = A.object $
       [ "action" A..= reqAction ]
    ++ [ "host_token"      A..= v | Just v <- [reqHostToken] ]
    ++ [ "host_token_kind" A..= v | Just v <- [reqHostTokenKind] ]
    ++ [ "target"          A..= v | Just v <- [reqTarget] ]
    ++ [ "trace_id"        A..= v | Just v <- [reqTraceId] ]

data DecideResponse = DecideResponse
  { respDecision   :: Decision
  , respReason     :: T.Text
  , respProofId    :: T.Text
  , respApprovalId :: Maybe T.Text
  , respDangerTags :: [T.Text]
  } deriving (Eq, Show)

instance A.FromJSON DecideResponse where
  parseJSON = AT.withObject "DecideResponse" $ \o -> do
    decTxt <- o .:? "decision" .!= ("unknown" :: T.Text)
    DecideResponse
      <$> pure (parseDecision decTxt)
      <*> o .:? "reason"      .!= ""
      <*> o .:? "proof_id"    .!= ""
      <*> o .:? "approval_id"
      <*> o .:? "danger_tags" .!= []

data TrustforgeError
  = DaemonUnavailable String
  | DaemonRejected   Int String
  | InvalidResponse  String
  deriving (Eq, Show)

data Client = Client
  { clientCfg     :: Config
  , clientManager :: H.Manager
  }

-- | Inspect the configuration the client was created with.
clientConfig :: Client -> Config
clientConfig = clientCfg

newClient :: Config -> IO Client
newClient cfg = do
  let timeoutMicros = max 1 (cfgTimeoutMs cfg) * 1000
      ms = H.defaultManagerSettings
        { H.managerResponseTimeout = H.responseTimeoutMicro timeoutMicros }
  mgr <- H.newManager ms
  pure (Client cfg mgr)

-- | Encode a 'DecideRequest' to a strict bytestring JSON body.
encodeRequestBody :: DecideRequest -> BS.ByteString
encodeRequestBody = LBS.toStrict . A.encode

-- | Decode the daemon's response body.
parseResponseBody :: LBS.ByteString -> Either TrustforgeError DecideResponse
parseResponseBody bs = case A.eitherDecode bs of
  Left e  -> Left (InvalidResponse e)
  Right r -> Right r

decide :: Client -> DecideRequest -> IO (Either TrustforgeError DecideResponse)
decide Client{clientCfg = cfg, clientManager = mgr} req = do
  let url = T.unpack (cfgDaemonUrl cfg) <> "/v1/decide"
  case H.parseRequest url of
    Left e -> pure (Left (DaemonUnavailable (show e)))
    Right base -> do
      let bearerHeaders =
            [ (CI.mk "authorization", BC.pack ("Bearer " <> T.unpack t))
            | Just t <- [cfgAdminToken cfg] ]
          httpReq = base
            { H.method        = HM.methodPost
            , H.requestBody   = H.RequestBodyBS (encodeRequestBody req)
            , H.requestHeaders =
                (CI.mk "content-type", "application/json") : bearerHeaders
            }
      result <- (Right <$> H.httpLbs httpReq mgr) `E.catch`
                  (\(e :: H.HttpException) ->
                      pure (Left (DaemonUnavailable (show e))))
      case result of
        Left err -> pure (Left err)
        Right resp ->
          let code = HS.statusCode (H.responseStatus resp)
              body = H.responseBody resp
          in if code >= 500
               then pure (Left (DaemonUnavailable ("status " <> show code)))
               else if code >= 400
                      then pure (Left (DaemonRejected code (show body)))
                      else pure (parseResponseBody body)

-- | Pull a Bearer token out of an Authorization header value.
extractBearer :: BS.ByteString -> Maybe T.Text
extractBearer h
  | BS.length h <= 7 = Nothing
  | BC.map toLower' (BC.take 7 h) /= "bearer " = Nothing
  | otherwise =
      let raw = BC.drop 7 h
          stripped = BC.dropWhile (== ' ') (BC.reverse (BC.dropWhile (== ' ') (BC.reverse raw)))
      in if BS.null stripped
           then Nothing
           else Just (TE.decodeUtf8 stripped)
  where
    toLower' c
      | c >= 'A' && c <= 'Z' = toEnum (fromEnum c + 32)
      | otherwise            = c
